src/clj/puppetlabs/services/request_handler/request_handler_core.clj

```diff
@@ -12,6 +12,14 @@
             [slingshot.slingshot :as sling]))
 
 ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;; Constants
+
+(def error-message-404
+  (str "Error: This master is running Puppet 4, and the URL format has changed. "
+       "This request was likely made with a Puppet 3 Agent, as it used an invalid "
+       "URL format. Please upgrade your agents to Puppet 4 and try again."))
```

Comment:
> So, the biggest thing I'd like to get in here is a block of server info, listing the software and API versions involved. That's probably the nicest and most useful 404 API response I can imagine.

Yes, this information is helpful so the end user knows where they stand presently and can plan a course of action to resolve the problem, for example, "I need to find the upgrade documentation for version X to version Y."